### PR TITLE
feat: streaming manifest_write for yaml-stream output

### PR DESCRIPTION
**Pull request by dasch**
## Summary

- Adds a `manifest_write` method to the `ManifestFormat` trait that writes output directly to an `io::Write` sink
- `YamlStreamFormat` overrides it to evaluate/manifest each array element individually, flushing to the writer and clearing the buffer between elements
- Memory usage for `--yaml-stream` is now O(largest single element) instead of O(total output)
- The CLI uses `manifest_write` for both stdout and file output paths, wrapped in `BufWriter`

## Motivation

When writing very large YAML streams (e.g. hundreds of thousands of Kubernetes manifests), the current implementation accumulates the entire manifested output into a single `String` before writing it. This defeats the lazy evaluation model of jsonnet — even though array elements are evaluated on-demand, the manifested strings are never freed.

With this change, each stream document is evaluated, manifested into a reusable buffer, flushed to the writer, and then the buffer is cleared before processing the next element.

## Design

The `ManifestFormat` trait gains a new method with a default implementation:

```rust
fn manifest_write(&self, val: Val, out: &mut dyn io::Write) -> Result<bool> { ... }
```

- Default: buffers into a String via `manifest()`, writes all at once (backward compatible)
- `YamlStreamFormat`: overrides to stream per-document
- Returns `bool` indicating whether any output was written (for trailing newline logic)

## Test plan

- [x] All existing tests pass (`cargo test -p jrsonnet -p jrsonnet-evaluator -p jrsonnet-stdlib -p tests`)
- [x] Manual verification of all output paths: stdout, file, yaml-stream, json, string, empty edge cases
- [x] Verified streaming works with `head` (broken pipe detected immediately, confirming incremental writes)

### CODE REVIEW

**Comment by dasch** [comment]
I think there's a similar issue and opportunity with `--multi`, but right now we're using `--yaml-stream`, so focusing on that first.

File: cmds/jrsonnet/src/main.rs [hunk 1]
```diff
@@ -249,13 +250,18 @@ fn main_real(opts: Opts) -> Result<(), Error> {
 			dir.pop();
 			create_dir_all(dir)?;
 		}
-		let mut file = File::create(path)?;
-		writeln!(file, "{}", val.manifest(manifest_format)?)?;
+		let mut file = std::io::BufWriter::new(File::create(path)?);
```
Comment (CertainLach) [inline cmds/jrsonnet/src/main.rs:253]: This leaves partially written files, but even if this concern is resolved, I do not think this will change anything

File: cmds/jrsonnet/src/main.rs [hunk 2]
```diff
@@ -249,13 +250,18 @@ fn main_real(opts: Opts) -> Result<(), Error> {
...
+		if manifest_format.manifest_write(val, &mut file)? {
+			file.write_all(b"\n")?;
+		}
+		file.flush()?;
 	} else {
-		let output = val.manifest(manifest_format)?;
-		if !output.is_empty() {
-			println!("{output}");
+		let stdout = std::io::stdout();
```
Comment (CertainLach) [inline cmds/jrsonnet/src/main.rs:259]: Partial writes are not the expected behavior for stdout either

**Comment by CertainLach** [comment]
I don't understand how that would help you with large manifests

As you hold ArrValue, its internal items would not be freed, so even if you buffer the output, it should not reduce memory usage at all, as the most of the memory use comes from the internal structs and their internal data, and not from the serialized buffers (unless you have a lot of duplication in the output, but I don't think even in this case you will gain anything meaningful from this change)

**Comment by dasch** [comment]
We have thousands of medium sized objects. So as it stands now, each item won't be freed – but would it not be possible to change that? I've used this pattern with Haskell in the past.

My assumption with `--yaml-stream` is that the output is a _stream_, not a list, and streams are not bounded in length, and aren't "complete".

**Comment by CertainLach** [comment]
> but would it not be possible to change that

Not in general with jsonnet.

If you have a jsonnet array - any item evaluated from it should be evaluated lazily and cached

It is in theory possible to extract the array data, drop the array, and then drop the array elements during manifestification - but that doesn't work in general either, because .jsonnet file imports are also cached, meaning you still have one reference to the underlying array in file evaluation cache. None of jsonnet implementations do that.

It is possible to process some data in streaming fashion instead, so there is no overarching structure that will keep all the elements in memory during manifestification, but there is no standard stdlib/jsonnet method that is able to do that, such things are usually achieved with the programmatic usage of jsonnet api.

Summarizing: in jsonnet, it is impossible to have proper output stream if the input is not a stream

**Comment by dasch** [comment]
Ah, I guess the fundamental problem is that jsonnet uses arrays rather than linked lists :/

I mean... it _could_ perhaps be done with a syntax extension, but maybe that would be less interesting? Basically introducing linked lists?

**Comment by CertainLach** [comment]
Linked lists won't help here

**Comment by CertainLach** [comment]
I don't think it is actionable/makes sense.
